### PR TITLE
Oak Graph view

### PR DESCRIPTION
**Pull request by jindraivanek**
Graph view for Oak tree.

image

Graph view can be enabled in settings, in Oak tab.

Rectangle nodes represents subtree, that was reduced to single node because of node limit.

Every node is clickable, then subtree from that node is viewed - back to upper level with "back" button.

TODO:
- [x] move "back" button to upper right corner

### CODE REVIEW

**Comment by jindraivanek** [comment]
demo
GraphView

File: src/client/fsharp/Oak/View.fs [hunk 1]
```diff
@@ -1,88 +1,318 @@
...
+let createGraph =
+    let getColor =
+        function
+        | Std -> "skyblue"
+        | Comment -> "lightgreen"
+        | Newline -> "gray"
+        | Directive -> "fuchsia"
+
+    memoizeBy fst
+    <| fun (model, dispatch: Msg -> unit) ->
```
Comment (nojaf) [inline src/client/fsharp/Oak/View.fs:215]: Nitpick but please refactor all code that uses a `<|`.

**Comment by jindraivanek** [comment]
added option for scaling size of nodes

**Review by nojaf — CHANGES_REQUESTED** [review]
First of amazing work! I'm really happy about this.

I must say I don't think the node structure is always correct:

When trying out:
```fsharp
module Tour.TuplesAndRecords

// From https://docs.microsoft.com/en-us/dotnet/fsharp/tour
// Visit the link above for more information on each topic
// You can also find more learning resources at https://fsharp.org/

module RecordTypes =

  /// This example shows how to define a new record type.
  type ContactCard =
    { Name: string;
      Phone: string;
      Verified: bool }

  /// This example shows how to instantiate a record type.
  let contact1 =
    { Name = "Alf";
      Phone = "(206) 555-0157";
      Verified = false }

  /// This example shows how to use "copy-and-update" on record values. It creates
  /// a new record value that is a copy of contact1, but has different values for
  /// the 'Phone' and 'Verified' fields.
  ///
  /// To learn more, see: https://docs.microsoft.com/dotnet/fsharp/language-reference/copy-and-update-record-expressions
  let contact2 =
    { contact1 with
        Phone = "(206) 555-0112";
        Verified = true }

  let ohSomeObjectExpression = { new System.Object() with member x.ToString() = "F#" }

  /// This example shows how to write a function that processes a record value.
  /// It converts a 'ContactCard' object to a string.
  let showContactCard (c: ContactCard) =
    c.Name
    + " Phone: "
    + c.Phone
    + (if not c.Verified then " (unverified)" else "")

  printfn "Alf's Contact Card: %s" (showContactCard contact1)
```

I got

image

The `TypeDefnRecordNode` should not have bindings as children.

And this probably is a small parsing bug but truth be told the client shouldn't really be parsing this back to a tree structure. As the server had the tree at some point.

I'd be more in favour of changing the backend a bit to return the tree as JSON and take it from there.

If update the backend, would you be willing to update the client?

File: src/client/fsharp/Oak/View.fs [hunk 2]
```diff
@@ -201,12 +201,22 @@ let limitTreeByNodes =
         limitTree allowedNodes n)
 
 let createGraph =
+    // copied from variables.sass
+    let colors = {|
```
Comment (nojaf) [inline src/client/fsharp/Oak/View.fs:205]: This is mind-blowingly clever to use an anonymous record at this location.
Pure genius 🤯

**Comment by jindraivanek** [comment]
> First of amazing work! I'm really happy about this.

Thanks!

> And this probably is a small parsing bug 

Indeed, I found out it was easy (one-char :) ) fix, so I did it.

>but truth be told the client shouldn't really be parsing this back to a tree structure. As the server had the tree at some point.
> 
> I'd be more in favour of changing the backend a bit to return the tree as JSON and take it from there.
> 
> If update the backend, would you be willing to update the client?

But I agree, getting tree from server would be better and more robust. Will be happy to update client.

File: src/client/fsharp/Oak/View.fs [hunk 3]
```diff
@@ -142,14 +142,11 @@ let rec private parseNode level (n: OakNode) =
 
 let private parseResults =
     memoize (fun (model: Model) ->
-        JS.console.log model.Oak
-
         if model.Oak = Unchecked.defaultof<_> then
```
Comment (nojaf) [inline src/client/fsharp/Oak/View.fs:145]: Feel free to change `model.Oak` to an option type if that makes things nicer.

**Comment by jindraivanek** [comment]
@nojaf Updated to consume tree from server. Not sure if Oak result is 100% same as before, because it is reconstructed after tree process.

**Comment by nojaf** [comment]
> Not sure if Oak result is 100% same as before, because it is reconstructed after tree process.

I'm afraid the outcome is still not correct indeed.

Would you mind extracting all the graph code to a separate file?
I don't think both views need to share any code.
For the default view, the JSON can just be transformed to `div`s as is.
That might be helpful in tracking down what is still going wrong.

Almost there!

**Comment by jindraivanek** [comment]
Moved code related to GraphView to separate file. I compared graph with Oak results in that @nojaf example, and looks like it is correct. 

There is slight difference thought, Trivia nodes are connected as childrens to node where they are as `ContentBefore/After`  , but in Oak results it looks like they are siblings. We can make different style of edge (and maybe reverse direction of arrow) to represent Trivia connection, to avoid confusion.

**Comment by nojaf** [comment]
I think it is correct now, one thing that still rubs me the wrong way is that the top-level node is not on top:

image

I passed the level property as well for this:

image

> There is slight difference thought, Trivia nodes are connected as childrens to node where they are as ContentBefore/After , but in Oak results it looks like they are siblings. We can make different style of edge (and maybe reverse direction of arrow) to represent Trivia connection, to avoid confusion.

Yes, good point. We could go all the way and implement our own [ctxRenderer](https://visjs.github.io/vis-network/docs/network/nodes.html). But that might be a stretch.
I think if we can create the trivia nodes as siblings nodes (with the same level property) and a different colour that might work.

**Comment by jindraivanek** [comment]
> one thing that still rubs me the wrong way is that the top-level node is not on top:

Great, thanks.

> Yes, good point. We could go all the way and implement our own [ctxRenderer](https://visjs.github.io/vis-network/docs/network/nodes.html). But that might be a stretch. I think if we can create the trivia nodes as siblings nodes (with the same level property) and a different colour that might work.

I think that's not needed, we can change style of edge, I quickly did one option:
image
If you are in favor of this, I'll do it properly.
For other options of edge styles, see https://visjs.github.io/vis-network/docs/network/edges.html

**Review by nojaf — APPROVED** [review]
Wonderful! Big up @jindraivanek 🚀

**Pull request merged**